File: Leetcode/arrays-and-strings/two-pointers/tasks/reverse-words-in-a-string-III/reverseWordsInAStringIII.ts
function reverseWords(s: string): string {
    let lastSpaceIndex = -1;
    const arrayOfChars = s.split('');

    for (let stringIndex = 0; stringIndex <= arrayOfChars.length; stringIndex++) {
        if (stringIndex === arrayOfChars.length || arrayOfChars[stringIndex] === ' ') {
            let startIndex = lastSpaceIndex + 1;
            let endIndex = stringIndex - 1;

            while (startIndex < endIndex) {
                const temp = arrayOfChars[startIndex];
                arrayOfChars[startIndex] = arrayOfChars[endIndex];
                arrayOfChars[endIndex] = temp;
                startIndex++;
                endIndex--;
            }

            lastSpaceIndex = stringIndex;
        }
    }

    return arrayOfChars.join('');
}

// Time: O(n + n) => O(n);
// Space: O(n);

const result = reverseWords('Let\'s take LeetCode contest');
console.log(result);

File: Leetcode/arrays-and-strings/two-pointers/tasks/reverse-words-in-a-string-III/reverseWordsInAStringIII.java
public String reverseWords(String s) {
    int lastSpaceIndex = -1;
    char[] arrayOfChars = s.toCharArray();

    for (int i = 0; i <= s.length(); i++) {
        if (i == s.length() || arrayOfChars[i] == ' ') {
            int startIndex = lastSpaceIndex + 1;
            int endIndex = i - 1;

            while (startIndex < endIndex) {
                char temp = arrayOfChars[startIndex];
                arrayOfChars[startIndex] = arrayOfChars[endIndex];
                arrayOfChars[endIndex] = temp;
                startIndex++;
                endIndex--;
            }

            lastSpaceIndex = i;
        }
    }

    return new String(arrayOfChars);
}

// Time: O(n + n) => O(n);
// Space: O(n);
